File: web/lib/types/inference-provider.ts
/**
 * The kind of backend a provider row talks to. Mirrors the backend `InferenceProviderKind` enum
 * (ReelForge.Shared/Data/Models/Enums.cs) exactly.
 *
 * `Anthropic` targets Claude through Anthropic's first-party Messages API. It supports the `Chat`
 * and `Vision` capabilities only — Anthropic exposes no speech-to-text API, so the backend rejects
 * an `Anthropic` + `Transcription` row with a 400. See docs/anthropic-provider.md.
 */
export type InferenceProviderKind = 'AzureOpenAI' | 'OpenAICompatible' | 'Anthropic';

/**
 * What an InferenceProvider row can be used for. Mirrors the backend `InferenceProviderCapability`
 * enum (ReelForge.Shared/Data/Models/Enums.cs) and `InferenceProviderResponse.Capability`
 * (Inference.Api/Controllers/Dto/InferenceProviderDtos.cs) exactly. `Chat`, `Transcription`, and
 * `Vision` (Phase 2 of video editing — shot captioning, see docs/video-editing.md) each participate
 * in their own independent "at most one default" constraint, so a chat-provider picker (e.g. the
 * per-agent override) must filter to `Chat` only — never merely "not Transcription", which would
 * incorrectly admit `Vision` rows too.
 */
export type InferenceProviderCapability = 'Chat' | 'Transcription' | 'Vision';

export interface InferenceProvider {
  id: string;
  name: string;
  kind: InferenceProviderKind;
  capability: InferenceProviderCapability;
  endpoint: string;
  modelName: string;
  hasApiKey: boolean;
  apiKeyLastFour: string | null;
  isDefault: boolean;
  isEnabled: boolean;
  timeoutSeconds: number | null;
  createdAt: string;
  updatedAt: string;
  lastTestAt: string | null;
  lastTestOk: boolean | null;
  lastTestError: string | null;
}

export interface CreateInferenceProviderRequest {
  name: string;
  kind: InferenceProviderKind;
  /** Omitted defaults to 'Chat' server-side, for backward compatibility. */
  capability?: InferenceProviderCapability;
  endpoint: string;
  modelName: string;
  apiKey?: string | null;
  isDefault: boolean;
  isEnabled: boolean;
  timeoutSeconds?: number | null;
}

export interface UpdateInferenceProviderRequest {
  name?: string;
  kind?: InferenceProviderKind;
  /** Omitted/null leaves the stored capability unchanged. */
  capability?: InferenceProviderCapability;
  endpoint?: string;
  modelName?: string;
  apiKey?: string | null;
  isDefault?: boolean;
  isEnabled?: boolean;
  timeoutSeconds?: number | null;
}

export interface TestInferenceProviderRequest {
  id?: string | null;
  kind?: InferenceProviderKind | null;
  capability?: InferenceProviderCapability | null;
  endpoint?: string | null;
  modelName?: string | null;
  apiKey?: string | null;
}

export interface TestInferenceProviderResponse {
  ok: boolean;
  latencyMs: number;
  error: string | null;
  responsePreview: string | null;
}

File: inference/src/ReelForge.Shared/Data/Models/InferenceProvider.cs
namespace ReelForge.Shared.Data.Models;

/// <summary>
/// A configured chat-completion inference backend (Azure OpenAI or an OpenAI-compatible
/// endpoint such as vLLM, LM Studio, llama.cpp server, or LiteLLM).
/// Exactly one row is expected to have <see cref="IsDefault"/> set to true.
/// </summary>
public class InferenceProvider
{
    public Guid Id { get; set; }

    /// <summary>Unique display name.</summary>
    public string Name { get; set; } = string.Empty;

    public InferenceProviderKind Kind { get; set; }

    /// <summary>Azure resource URL, or OpenAI-compatible base URL.</summary>
    public string Endpoint { get; set; } = string.Empty;

    /// <summary>Azure deployment name, or model id.</summary>
    public string ModelName { get; set; } = string.Empty;

    /// <summary>Data-Protection ciphertext; never leaves the backend.</summary>
    public string? ApiKeyEncrypted { get; set; }

    /// <summary>Display-only hint, e.g. the last four characters of the key.</summary>
    public string? ApiKeyLastFour { get; set; }

    public bool IsDefault { get; set; }

    public bool IsEnabled { get; set; } = true;

    public int? TimeoutSeconds { get; set; }

    /// <summary>jsonb; some self-hosted gateways need a custom auth header.</summary>
    public string? ExtraHeadersJson { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    public DateTime? LastTestAt { get; set; }

    public bool? LastTestOk { get; set; }

    public string? LastTestError { get; set; }

    public ICollection<AgentDefinition> AgentDefinitions { get; set; } = new List<AgentDefinition>();
}
